---
title: "CAS_scRNAseq_analysis"
author: "James Read, PhD"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



### Load libraries

```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=TRUE}
library(Matrix)
library(DropletUtils)
library(scater)
library(SingleCellExperiment)
library(Seurat)
library(org.Hs.eg.db) # Human
library(cowplot)
library(dplyr)
library(biomaRt)
library(DoubletFinder)
library(SingleR)
library(gridExtra)
library(tictoc)
```

```{r, warning=FALSE, message=FALSE, echo=TRUE, eval=FALSE}
library(Matrix)
library(DropletUtils)
library(scater)
library(SingleCellExperiment)
library(Seurat)
library(org.Hs.eg.db) # Human
library(cowplot)
library(dplyr)
library(biomaRt)
library(DoubletFinder)
library(SingleR)
library(gridExtra)
library(tictoc)
```

```{r, warning=FALSE, echo=FALSE, eval=TRUE}
In.dir = "C:/Users/jread/Desktop/Work/Git/CAS_scRNAseq_analysis/Raw_data"
Out.dir1 = "C:/Users/jread/Desktop/Work/Git/CAS_scRNAseq_analysis/QC"
Out.dir2 = "C:/Users/jread/Desktop/Work/Git/CAS_scRNAseq_analysis/Output"
```


```{r, warning=FALSE, echo=TRUE, eval=FALSE}
In.dir = "Path/To/Raw_data"
Out.dir1 = "Path/To/QC"
Out.dir2 = "Path/To/Output"
```


# Part 1 - QC and Preprocessing

## Preprocessing function

This function applies preprocessing to raw data indicated in the function input. The outputs will be saved in the 'QC' directory
```{r, warning=FALSE, message=FALSE}
preprocess_function = function(d, age, stim, save.obj) {
  
  ##### Setup #####
  print("Setting up...")
  nam = paste(d, age, stim, sep = "_")
  
  x1 <- readMM(paste0(nam, "_matrix.mtx.gz"))
  Barcodes <- read.table(paste0(nam, "_barcodes.tsv.gz"))
  Genes <- read.table(paste0(nam, "_features.tsv.gz"))
  rownames(x1) = Genes$V1; colnames(x1) = Barcodes$V1
  
  # createt output directory
  setwd(Out.dir1)
  dir.create(nam)
  setwd(paste("./", nam, sep=""))
  
  # Remove empty droplets
  print("Removing empty droplets...")
  bcrank = barcodeRanks(x1)
  uniq = !duplicated(bcrank$rank)
  
  # Save plot
  png(paste("1)", nam, "BarcodeRanks.png", sep="_"), units="in", width=5, height=5, res=300)
  invisible(plot(bcrank$rank[uniq], bcrank$total[uniq], main = nam, log="xy", xlab="Rank",
                 ylab="Total UMI count", cex.lab=1.2))
  abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
  abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)
  legend("bottomleft", legend = c("Knee", "inflection"), col=c("dodgerblue", "darkgreen"),
         lty=2, cex=1.2)
  invisible(dev.off())
  
  # Total droplets
  set.seed(12345) # Reproducibility
  e.out <- emptyDrops(x1, lower = 100, niters = 10000) # identify empty drops
  #head(e.out[order(-e.out[,1]),])
  #sum(e.out$FDR <=0.01, na.rm=TRUE)
  #table(Sig=e.out$FDR <= 0.01, Limited=e.out$Limited)
  x1 <- x1[,which(e.out$FDR<=0.01)] # remove 'empty drops'
  # Number not empty
  dim(x1)

  #### Filter genes ####
  print("Filtering genes...")
  # Create singleCellExperiment object
  sce <- SingleCellExperiment(assays = list(counts = as.matrix(x1)))
  colData(sce)$Barcode <- colnames(x1)
  rowData(sce)$EnsemblID <- rownames(x1)
  rowData(sce)$GeneID <- Genes$V2
  # Make sure rownames are unique
  rownames(sce) <- uniquifyFeatureNames(rowData(sce)$EnsemblID, rowData(sce)$GeneID)
  
  # Remove genes with no expression in any cells
  keep <- rowSums(counts(sce)>0)>0
  sce <- sce[keep,]
  
  # Retain only gene with expression in at least 1% of cells
  ncol(sce)*0.001
  keep_feature <- nexprs(sce, byrow = TRUE) >= ncol(sce)*0.001
  # Number to keep
  table(keep_feature)
  sce <- sce[keep_feature,]
  dim(sce)
  
  ##### Quality control of cells #####
  print("Quality control of cells...")
  
  # Map chromosome info
  map <- invisible(mapIds(org.Hs.eg.db, keys = as.character(rowData(sce)$EnsemblID),
                          column = "CHR", keytype = "ENSEMBL"))
  rowData(sce)$CHR <- map
  summary(map=="MT")
  
  p1 = data.frame(table(rowData(sce)$CHR))
  p1 = p1[order(as.numeric(-p1$Freq)),]
  
  png(paste("2)", nam, "GeneByChr.png", sep="_"), units="in", width=12, height=5, res=300)
  barplot(p1$Freq, main = paste(nam,"\nFeatures per chromosome"), xlab = "Chromosome", cex.main = 1.5,
                  col = alpha("cyan", 0.2), border = "cyan3", lwd = 3, names.arg = p1$Var1)
  invisible(dev.off())
  
  ## QC metrics
  # Get mitochondrial and ribosomal gene content
  mt_genes <- grep("^MT-", rownames(sce), ignore.case = TRUE, value = TRUE)
  rb_genes <- grep("^RPS|^RPL", rownames(sce), ignore.case = TRUE, value = TRUE)
  # Calculate QC metrics
  controls <- list(Mito = mt_genes, Ribo = rb_genes)
  qc <- invisible(addPerCellQC(sce, subsets=controls, rownames(sce)))
  
  # Plot by feature/transctipt per cell
  png(paste("3)", nam, "Log10Plots.png", sep="_"), units="in", width=12, height=5, res=300)
  par(mfrow=c(1,2))           
  hist(log(qc$detected, 10), main = paste(nam, "- No. features (Log10)\n[Number of unique genes per cell]"),
       xlab = "Log10 Unique Features", breaks = 50, col = "cyan", border = "grey")
  abline(v = c(log(250,10),log(500,10),log(1000,10),log(2000,10)), col = "black", lty = 2, lwd = 1.5)
  text(log(250,10),25, "250", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(500,10),50, "500", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(1000,10),75, "1000", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(2000,10),100, "2000", col = alpha("red", 0.5), adj = 1, cex=0.6)
  hist(log(qc$total, 10), main = paste(nam, "- Library size (Log10)\n[Number of transcripts per cell]"),
       xlab = "Log10 Total counts", breaks = 50, col = "cyan", border = "grey")
  abline(v = c(log(250,10),log(500,10),log(1000,10),log(2000,10)), col = "black", lty = 2, lwd = 1.5)
  text(log(250,10),25, "250", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(500,10),50, "500", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(1000,10),75, "1000", col = alpha("red", 0.5), adj = 1, cex=0.6)
  text(log(2000, 10),100, "2000", col = alpha("red", 0.5), adj = 1, cex=0.6)
  invisible(dev.off())
  
  # Retain if greater than 500 features (genes) or 1000 transcript
  keep.total <- qc$total > 500
  keep.n <- qc$detected > 1000
  table(keep.total); table(keep.n)#; table(keep.mt); table(keep.rb)
  qc.f <- qc[,keep.total & keep.n]
  dim(qc.f)
  
  # Remove cells with low feature to count ratio
  # NOTE: This step is outdated and unnecessary, included here for reproduciability
  qc.f$nFeature_nCount_ratio = qc.f$detected/qc.f$total
  png(paste("4)", nam, "FeatureCountRatio.png", sep="_"), units="in", width=5, height=5, res=300)
  plot(qc.f$nFeature_nCount_ratio, ylab = "Feature/count", pch = 21, col = "black",
            bg = alpha("grey50", 0.5), main = paste(nam, "\nFeature to count ratio"), cex = 0.7)
  abline(h = 0.1, col = "cyan3", lty = 2, lwd=2)
  if (summary(qc.f$nFeature_nCount_ratio)[[1]] < 0.01) {
    legend("bottomright", legend = paste(table(qc.f$nFeature_nCount_ratio<0.1)[[2]], "cells removed"),
         text.font = 3, bty='n', text.col = "grey50", cex=0.7)
  }
  
  if (summary(qc.f$nFeature_nCount_ratio)[[1]] > 0.01) {
    legend("bottomright", legend = "0 cells removed",
         text.font = 3, bty='n', text.col = "grey50", cex=0.7)
  }
  
  invisible(dev.off())
  # Remove cells if feature to count ratio < 0.1
  
  if (summary(qc.f$nFeature_nCount_ratio)[[1]] < 0.01) {
    qc.f = qc.f[, !qc.f$nFeature_nCount_ratio<0.1]
  }
  
    # Plot by mitochondrial content
  ## Using a threshold of 3x the Median Absolute Deviation (MAD) as high mitochondrial content
  q.met = qc.f$subsets_Mito_percent
  png(paste("5)", nam, "Mito_Ribo_Plots.png", sep="_"), units="in", width=12, height=4, res=300)
  par(mfrow = c(1, 3))
  plot(log10(qc.f$total), q.met, xlab = "Log10(Total genes)", ylab = "Mitochondrial genes (%)",
            pch = 21, col = "black", bg = alpha("grey50", 0.5),
            main = paste(nam, "\nMitochondrial content"), cex = 0.7)
  abline(h = median(q.met) + 3*mad(q.met), col = "dodgerblue", lty = 2, lwd=2)
  legend("topright", legend = paste("Mitochondrial gene threshold = ",
                                    round(median(q.met) + 3*mad(q.met),2), "%", sep=""),
         bty='n', text.col = "dodgerblue2", cex=0.7)
  plot(log10(qc.f$total), qc.f$subsets_Ribo_percent, xlab = "Log10(Total genes)",
       ylab = "Ribosomal RNA (%)", pch = 21, col = "black", bg = alpha("grey50", 0.5),
       main = paste(nam, "\nRibosomal content"), cex = 0.7)
  abline(h = 50, col = "darkorchid1", lty = 2, lwd=2)
  legend("topright", legend = "Ribosomal gene threshold = 50%", bty='n',
         text.col = "darkorchid2", cex=0.7)
  plot(qc.f$subsets_Ribo_percent, log(q.met), ylab = "Log[Mitochondrial genes (%)]",
       xlab = "Ribosomal RNA (%)", pch = 21, col = "black", bg = alpha("grey50", 0.5),
       main = paste(nam, "\nMitochondrial vs Ribosomal content"), cex = 0.7)
  abline(h = log(median(q.met) + 3*mad(q.met)), col = "dodgerblue", lty = 2, lwd=2)
  abline(v = 50, col = "darkorchid1", lty = 2, lwd=2)
  polygon(x = c(-5, max(qc.f$subsets_Ribo_percent)+5, max(qc.f$subsets_Ribo_percent)+5, -5),
          y=c(log(median(q.met) + 3*mad(q.met)), log(median(q.met) + 3*mad(q.met)), max(log(q.met)+1),
              max(log(q.met)+1)), col = alpha("grey", 0.3), border = alpha("grey", 0.3))
  polygon(x = c(50, max(qc.f$subsets_Ribo_percent)+5, max(qc.f$subsets_Ribo_percent)+5, 50),
          y=c(-5, -5, max(log(q.met)+1), max(log(q.met)+1)), col = alpha("grey", 0.3),
          border = alpha("grey", 0.3))
  in.out = table(q.met < (median(q.met) + 3*mad(q.met)) & qc.f$subsets_Ribo_percent < 50)
  legend("top", legend = paste("Excluded (n=", in.out[[1]], ")",sep=""), bty='n',
         text.col = "grey30", cex=0.7, text.font=2)
  legend("bottomleft", legend = paste("Retained (n=", in.out[[2]], ")",sep=""), bty='n',
         text.col = "navy", cex=0.7, text.font=2)
  invisible(dev.off())
  
  # Remove cells with mitochondrial content above 3x the MAD or ribosomal content greater than 50%
  sce1 = qc.f[, (q.met < (median(q.met) + 3*mad(q.met)) & qc.f$subsets_Ribo_percent < 50)]
  
  # Plot after filtering cells
  png(paste("6)", nam, "Log10Plots_AfterQC1.png", sep="_"), units="in", width=12, height=5, res=300)
  par(mfrow=c(1,2))
  hist(log(sce1$detected, 10),
            main = paste(nam, "- No. features (Log10)\n[Number of unique genes per cell]"),
       xlab = "Log10 Unique Features",
       breaks = 50, col = "cyan", border = "grey")
  hist(log(sce1$total, 10),
       main = paste(nam, "- Library size (Log10)\n[Number of transcripts per cell]"),
       xlab = "Log10 Total counts", breaks = 50, col = "cyan", border = "grey")
  invisible(dev.off())
  
  png(paste("7)", nam, "BeforeAfterQC.png", sep="_"), units="in", width=10, height=3.5, res=300)
  print(plot_grid(plotColData(qc, x = "detected", y="subsets_Mito_percent",
                              colour_by="subsets_Ribo_percent") +
                    ggtitle(nam) + scale_x_log10(limits = c(50,10000)) + 
                    ylim(0,100) + stat_density2d(colour = "black", bins = 10),
                  plotColData(sce1, x = "detected", y="subsets_Mito_percent",
                              colour_by="subsets_Ribo_percent") + ggtitle(nam) +
                    scale_x_log10(limits = c(50,10000)) + ylim(0,100) + 
                    stat_density2d(colour = "black", bins = 10), ncol = 2))
  invisible(dev.off())
  
  ##### Basic Seurat processing #####
  # create Seurat object
  print("Starting basic Seurat analysis...")
  S1 <- invisible(CreateSeuratObject(counts = assays(sce1)$counts,
                                     min.cells = 0, min.features = 0, project = nam))
  S1@meta.data$percent.mt <- colData(sce1)$subsets_Mito_percent
  S1@meta.data$percent.rb <- colData(sce1)$subsets_Ribo_percent
  S1@misc = list(rowData(sce1)$EnsemblID)
  
  # Preprocess
  print("Normalizing....")
  S1 <- NormalizeData(S1, normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE)
  S1 <- FindVariableFeatures(S1, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  top10 <- head(VariableFeatures(S1), 10)
  ## Plot variable features
  png(paste("8)", nam, "VarGenesPlot.png", sep="_"), units="in", width=7.5, height=4, res=300)
  print(invisible(LabelPoints(plot = VariableFeaturePlot(S1), points = top10, repel = TRUE)) +
          ggtitle(nam))
  invisible(dev.off())
  
  # Scale and PCA
  print("Running PCA...")
  S1 <- ScaleData(S1, features = rownames(S1), verbose = FALSE)
  S1 <- RunPCA(S1, features = VariableFeatures(object = S1), verbose = FALSE)
  # Plot loadings of PC 1&2
  png(paste("9)", nam, "PCSLoadingsPlot.png", sep="_"), units="in", width=7.5, height=5, res=300)
  print(VizDimLoadings(S1, dims = 1:2, reduction = "pca") + ggtitle(nam))
  invisible(dev.off())
  # Plot PCA
  png(paste("10)", nam, "PCAPlot.png", sep="_"), units="in", width=4, height=4, res=300)
  print(DimPlot(S1, reduction = "pca") + NoLegend() + ggtitle(nam))
  invisible(dev.off())
  # Heatmap of PC1
  png(paste("11)", nam, "HeatmapPlot_PC1.png", sep="_"), units="in", width=5, height=4, res=300)
  print(DimHeatmap(S1, dims = 1, cells = 500, balanced = TRUE, fast = F) + ggtitle(nam))
  invisible(dev.off())
  # Plot heatmap of first 9 PCs
  png(paste("12)", nam, "HeatmapPlot_9PC.png", sep="_"), units="in", width=10, height=10, res=300)
  DimHeatmap(S1, dims = 1:9, cells = 500, balanced = TRUE)
  invisible(dev.off())
  
  # Jackstraw for PC significance
  S1 <- JackStraw(S1, num.replicate = 100, verbose = FALSE)
  S1 <- ScoreJackStraw(S1, dims = 1:20)
  png(paste("13)", nam, "JackStraw.png", sep="_"), units="in", width=5, height=5, res=300)
  print(invisible(JackStrawPlot(S1, dims = 1:20)) + ggtitle(nam))
  invisible(dev.off())
  # Jackstraw elbow plot
  png(paste("14)", nam, "JSElbowPlot.png", sep="_"), units="in", width=5, height=5, res=300)
  print(ElbowPlot(S1) + ggtitle(nam))
  invisible(dev.off())
  
  #### UMAP ####
  print("Running UMAP analysis...")
  # Clustering
  S1 <- FindNeighbors(S1, dims = 1:15, verbose = FALSE)
  S1 <- FindClusters(S1, resolution = 0.2, verbose = FALSE)
  # Run UMAP
  S1 <- invisible(RunUMAP(S1, dims = 1:15, verbose = FALSE))
  png(paste("15)", nam, "UMAP_clusters.png", sep="_"), units="in", width=5, height=4, res=300)
  print(DimPlot(S1, reduction = "umap") + ggtitle(nam))
  invisible(dev.off())
  # Find cluster markers
  S1.markers <- FindAllMarkers(S1, only.pos = TRUE, min.pct = 0.25,
                               logfc.threshold = 0.25, verbose = FALSE)
  # Plot heatmap by cluster-determoining genes
  top10 <- S1.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
  png(paste("16)", nam, "Cluster_HM.png", sep="_"), units="in", width=8, height=8, res=300)
  print(DoHeatmap(S1, features = top10$gene, size = 3) + ggtitle(nam))
  invisible(dev.off())

  # Cell cycle
  print("Calculating cell cycle phase...")
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  # Estimate cell cycle phase
  S1 = invisible(CellCycleScoring(S1, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE))
  # Plot over UMAP
  png(paste("17)", nam, "CellCyclePlot.png", sep="_"), units="in", width=5, height=4, res=300)
  print(DimPlot(S1, reduction = "umap", group.by = "Phase") + ggtitle(nam))
  invisible(dev.off())
  # Plot Barplot by cell cycle phase
  test1 = data.frame(S1@meta.data$seurat_clusters, S1@meta.data$Phase)
  test2 = table(test1$S1.meta.data.seurat_clusters, test1$S1.meta.data.Phase)
  png(paste("18)", nam, "CellCycleBar.png", sep="_"), units="in", width=5, height=4.5, res=300)
  barplot(t(test2), col = c("#F8766D", "#0CB702", "#00A9FF"), xlab="Cluster",
          ylab = "Number of cells", main = paste(nam, "\n Cell cycle phase by cluster"),
          cex.lab=1.5)
  legend("topright", legend = c("G1", "G2M", "S1"), col=c("#F8766D", "#0CB702", "#00A9FF"),
         pch=15, cex=1.2)
  invisible(dev.off())

  ##### Doublet detection #####
  print("Detecting doublets...")
  hush=function(code){
    sink("NUL")
    tmp = code
    sink()
    return(tmp)
    } # supress some of the outputs
  sweep.res = hush(paramSweep(S1, PCs = 1:15, sct = FALSE))
  sweep.stats = summarizeSweep(sweep.res, GT = FALSE)
  # Detemine pK value
  bc.mvn <- as.data.frame(matrix(0L, nrow = length(unique(sweep.stats$pK)), ncol = 5))
  colnames(bc.mvn) <- c("ParamID", "pK", "MeanBC", "VarBC", "BCmetric")
  bc.mvn$pK <- unique(sweep.stats$pK); bc.mvn$ParamID <- 1:nrow(bc.mvn)
  x <- 0
  for (i in unique(bc.mvn$pK)) {
      x <- x + 1
      ind <- which(sweep.stats$pK == i)
      bc.mvn$MeanBC[x] <- mean(sweep.stats[ind, "BCreal"])
      bc.mvn$VarBC[x] <- sd(sweep.stats[ind, "BCreal"])^2
      bc.mvn$BCmetric[x] <- mean(sweep.stats[ind, "BCreal"])/(sd(sweep.stats[ind, "BCreal"])^2)}
  bc.mvn$pK_num = c(0.005, seq(0.01, 0.3, 0.01))
  opt.pK = bc.mvn[which(bc.mvn$BCmetric == max(bc.mvn$BCmetric)), "pK_num"]
  # estimate the homotypic proportion
  homotypic.prop = sum((table(Idents(S1))/sum(table(Idents(S1))))^2)
  nExp_poi <- round(0.05*ncol(S1))  ## Assuming 5% doublet rate from all cells
  nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
  # Plot optimum pK value
  png(paste("19)", nam, "Doublet_pK_optimum.png", sep="_"), units="in", width=5, height=5, res=300)
  plot(bc.mvn$pK_num, log(bc.mvn$BCmetric), ylab = "Log(metric)", xlab = "pK",
  main = "Mean-variance normalized\nbimodality coefficient (log)", pch = 19,
  type = 'b', col="dodgerblue2")
  abline(v=opt.pK, col=alpha("firebrick1", 0.5), lwd=3, lty=2)
  legend("topright", legend = paste("Optimum pK =", opt.pK), bty="n",
         text.font = 2, text.col = alpha("firebrick1", 0.5))
  invisible(dev.off())
  # Run DoubletFinder on Seurat object
  seu1 <- hush(doubletFinder(S1, PCs = 1:15, pN = 0.25, pK = opt.pK,
                           nExp = nExp_poi.adj, reuse.pANN = FALSE, sct = FALSE))
  #Add into orignial Seurat object
  identical(colnames(S1), colnames(seu1)) # Check objects match
  S1$Doublet = seu1@meta.data[, ncol(seu1@meta.data)]
  # Number of doublets
  table(S1$Doublet)
  # Plot Doublets over UMAP
  png(paste("20)", nam, "UMAP_doublets.png", sep="_"), units="in", width=5, height=4, res=300)
  print(DimPlot(S1, reduction = "umap", group.by = "Doublet") + ggtitle(nam) + 
          scale_color_manual(values=c("black", alpha("dodgerblue1", 0.15))))
  invisible(dev.off())
  # Remove doublets
  S1 = subset(S1, Doublet == "Singlet")
  
  ##### SingleR #####
  print("Broad annotation with SingleR...")
  # Using singleR to get an over view of cell types present
  # Get references
  ref.HPCA = invisible(HumanPrimaryCellAtlasData())
  ref.BPE = invisible(BlueprintEncodeData())
  # Setup data
  dat = as.matrix(S1@assays$RNA$data)
  clust = as.character(S1@meta.data$seurat_clusters)
  # Run singleR on HPCA and BPE references
  pred.HPCA <- SingleR(test = dat, ref = ref.HPCA, labels = ref.HPCA$label.main)
  pred.BPE <- SingleR(test = dat, ref = ref.BPE, labels = ref.BPE$label.main)
  S1@meta.data$SingleR_HPCA = pred.HPCA$pruned.labels
  S1@meta.data$SingleR_BPE = pred.BPE$pruned.labels
  # PLot HPCA annotation over UMAP
  png(paste("21)", nam, "SingleR_HPCA_UMAP.png", sep="_"), units="in", width=6, height=5, res=300)
  print(DimPlot(S1, reduction = "umap", group.by = "SingleR_HPCA") + ggtitle(nam))
  invisible(dev.off())
  # PLot BPE annotation over UMAP
  png(paste("22)", nam, "SingleR_BPE_UMAP.png", sep="_"), units="in", width=6, height=5, res=300)
  print(DimPlot(S1, reduction = "umap", group.by = "SingleR_BPE") + ggtitle(nam))
  invisible(dev.off())
  
  #### QC Metrics ####
  print("Tabulating QC metrics")
  # Tabulate some of the major QC metrics 
  ncol(S1) # Number of cells
  nrow(S1) # Number of genes detected
  sum(S1@assays$RNA$counts) # Total counts measured
  q1 = summary(colSums(S1@assays$RNA$counts))[[3]]; q1 # Median counts per cell
  q2 = summary(apply(S1@assays$RNA$counts, 2, function(x) sum(x>0)))[[3]]; q2 # Median genes per cell
  q3 = summary(S1@meta.data$percent.mt); q3 # Summary of Mitochodrial gene percentage per cell
  q4 = summary(S1@meta.data$percent.rb); q4 # Summary of Ribosomal gene percentage per cell
  # Plot QC metrics
  out.tab = data.frame(ncol(S1), nrow(S1), sum(S1@assays$RNA$counts), q1, q2,
                       round(q3[[3]], 3), round(q3[[6]], 3), round(q4[[3]], 3))
  colnames(out.tab) = c('No. cells', 'No. genes', 'Total counts', 'Median counts/cell',
                        'Median genes/cell', "Median mitochodrial %", 
                        "Maximum mitochodrial %", "Median ribosomal %")
  rownames(out.tab) = nam
  # Plot table
  ss <- tableGrob(t(out.tab))
  png(paste("23)", nam, "QC_metric_summary.png", sep="_"), units="in", width=5, height=5, res=300)
  grid.arrange(ss)
  invisible(dev.off())
  
  print("Saving data....")
  # Save preprocessed Seurat object
  S1$Sample= d
  S1$Age= age
  S1$Stimuli= stim
  S1$Group = paste(S1$Sample, S1$Age, S1$Stimuli, sep = "_")
  if (save.obj == "TRUE") {
    saveRDS(S1, paste(nam,"_cleaned.rds", sep=""))
  }
  
  # Return preprocessed seurat object
  return(S1)
  
  print("Finished!!!!")
}
```

## Run the function

See raw files available for preprocessing. These are in the 'Raw_data' directory and were taken from the GEO repository (GSE232186) at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE232186
```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
list.files()
```

Now run the function to apply preprocessing to the raw data. The function will create several output plots and save a seurat object of the preprocessed data (if save.obj = "TRUE)

Note: 'save.obj' is set to 'FALSE' as the seurat object saved is > 100mb for each sample, which is larger than github allows.


## Donor 1 - Cord blood

### Donor 1 - Cord - Unstimulated

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic() # To time run
D1_Cord_Unstim = preprocess_function(d = "Donor1", age = "Cord", stim = "Unstim", save.obj = "FALSE")
toc()
# See output
class(D1_Cord_Unstim)
dim(D1_Cord_Unstim)
```

### Donor 1 - Cord - LPS

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D1_Cord_LPS = preprocess_function(d = "Donor1", age = "Cord", stim = "LPS", save.obj = "FALSE")
toc()
```

### Donor 1 - Cord - PolyIC

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D1_Cord_PIC = preprocess_function(d = "Donor1", age = "Cord", stim = "PIC", save.obj = "FALSE")
toc()
```


## Donor 1 - 5 year blood

### Donor 1 - 5yr - Unstimulated

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D1_5yr_Unstim = preprocess_function(d = "Donor1", age = "5yr", stim = "Unstim", save.obj = "FALSE")
toc()
```

### Donor 1 - 5yr - LPS

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D1_5yr_LPS = preprocess_function(d = "Donor1", age = "5yr", stim = "LPS", save.obj = "FALSE")
toc()
```

### Donor 1 - 5yr - PolyIC

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D1_5yr_PIC = preprocess_function(d = "Donor1", age = "5yr", stim = "PIC", save.obj = "FALSE")
toc()
```



## Donor 2 - Cord blood

### Donor 2 - Cord - Unstimulated

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_Cord_Unstim = preprocess_function(d = "Donor2", age = "Cord", stim = "Unstim", save.obj = "FALSE")
toc()
```

### Donor 2 - Cord - LPS

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_Cord_LPS = preprocess_function(d = "Donor2", age = "Cord", stim = "LPS", save.obj = "FALSE")
toc()
```

### Donor 2 - Cord - PolyIC

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_Cord_PIC = preprocess_function(d = "Donor2", age = "Cord", stim = "PIC", save.obj = "FALSE")
toc()
```


## Donor 2 - 5 year blood

### Donor 2 - 5yr - Unstimulated

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_5yr_Unstim = preprocess_function(d = "Donor2", age = "5yr", stim = "Unstim", save.obj = "FALSE")
toc()
```

### Donor 2 - 5yr - LPS

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_5yr_LPS = preprocess_function(d = "Donor2", age = "5yr", stim = "LPS", save.obj = "FALSE")
toc()
```

### Donor 2 - 5yr - PolyIC

```{r, warning=FALSE, message=FALSE}
setwd(In.dir)
tic()
D2_5yr_PIC = preprocess_function(d = "Donor2", age = "5yr", stim = "PIC", save.obj = "FALSE")
toc()
```


###########################################################

# Part 2 - Integration and Annotation


### Load libraries

```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=TRUE}
library(Seurat)
library(RColorBrewer)
library(pheatmap)
library(org.Hs.eg.db)
library(WGCNA)
library(scales)
library(ggplot2)
library(basicPlotteR)
library(enrichR)
library(tictoc)
```

```{r, warning=FALSE, message=FALSE, echo=TRUE, eval=FALSE}
library(Seurat)
library(RColorBrewer)
library(pheatmap)
library(org.Hs.eg.db)
library(WGCNA)
library(scales)
library(ggplot2)
library(basicPlotteR)
library(enrichR)
library(tictoc)
```


## Integration

```{r, warning=FALSE, message=FALSE, eval=TRUE, echo=FALSE}

setwd("C:/Users/jread/Desktop/Work/CAS_scRNAseq/New_Cleaned_data")

D1_Cord_Unstim = readRDS("Donor1_Cord_Unstim_cleaned.rds")
D1_Cord_LPS = readRDS("Donor1_Cord_LPS_cleaned.rds")
D1_Cord_PIC = readRDS("Donor1_Cord_PIC_cleaned.rds")
D1_5yr_Unstim = readRDS("Donor1_5yr_Unstim_cleaned.rds")
D1_5yr_LPS = readRDS("Donor1_5yr_LPS_cleaned.rds")
D1_5yr_PIC = readRDS("Donor1_5yr_PIC_cleaned.rds")
D2_Cord_Unstim = readRDS("Donor2_Cord_Unstim_cleaned.rds")
D2_Cord_LPS = readRDS("Donor2_Cord_LPS_cleaned.rds")
D2_Cord_PIC = readRDS("Donor2_Cord_PIC_cleaned.rds")
D2_5yr_Unstim = readRDS("Donor2_5yr_Unstim_cleaned.rds")
D2_5yr_LPS = readRDS("Donor2_5yr_LPS_cleaned.rds")
D2_5yr_PIC = readRDS("Donor2_5yr_PIC_cleaned.rds")

```

Make integration list
```{r}
s.obj.list = list(
  D1_Cord_Unstim = D1_Cord_Unstim,
  D1_Cord_LPS = D1_Cord_LPS,
  D1_Cord_PIC = D1_Cord_PIC,
  D1_5yr_Unstim = D1_5yr_Unstim,
  D1_5yr_LPS = D1_5yr_LPS,
  D1_5yr_PIC = D1_5yr_PIC,
  D2_Cord_Unstim = D2_Cord_Unstim,
  D2_Cord_LPS = D2_Cord_LPS,
  D2_Cord_PIC = D2_Cord_PIC,
  D2_5yr_Unstim = D2_5yr_Unstim,
  D2_5yr_LPS = D2_5yr_LPS,
  D2_5yr_PIC = D2_5yr_PIC
)
```

Select features
```{r}
features = SelectIntegrationFeatures(object.list = s.obj.list)
```

Find anchors
```{r, warning=FALSE, message=FALSE, eval = TRUE, echo=FALSE}
tic()
anchors = invisible(FindIntegrationAnchors(object.list = s.obj.list, k.anchor = 5,
                                           anchor.features = features,
                                           reduction = "rpca"))
toc()
```

```{r, warning=FALSE, message=FALSE, eval = FALSE, echo=TRUE}
tic()
anchors = FindIntegrationAnchors(object.list = s.obj.list, k.anchor = 5,
                                 anchor.features = features,
                                 reduction = "rpca")
toc()
```


Integrate data
```{r, warning=FALSE, message=FALSE, eval = TRUE, echo=FALSE}
tic()
s.obj = invisible(IntegrateData(anchorset = anchors))
toc()
```

Integrate data
```{r, warning=FALSE, message=FALSE, eval = FALSE, echo=TRUE}
tic()
s.obj = IntegrateData(anchorset = anchors)
toc()
```

Rejoin data layers
```{r}
s.obj[["RNA"]] <- JoinLayers(s.obj[["RNA"]])
```

Clean up
```{r}
rm(D1_Cord_LPS, D1_Cord_PIC, D1_Cord_Unstim, D1_5yr_LPS, D1_5yr_PIC,
   D1_5yr_Unstim, D2_Cord_LPS, D2_Cord_PIC, D2_Cord_Unstim,
   D2_5yr_LPS, D2_5yr_PIC, D2_5yr_Unstim, s.obj.list, anchors)
gc()
```


### Seurat dimensionality reduction and clustering

```{r, warning=FALSE, message=FALSE}
DefaultAssay(object = s.obj) <- "integrated"

s.obj <- ScaleData(s.obj)
s.obj <- RunPCA(s.obj, npcs = 30, verbose = FALSE)
s.obj <- RunUMAP(s.obj, reduction = "pca", dims = 1:30, 
                 n.neighbors = 20, min.dist = 0.3, verbose = FALSE)
s.obj <- FindNeighbors(s.obj, reduction = "pca", dims = 1:30, verbose = FALSE)
s.obj <- FindClusters(s.obj, resolution = 0.4, verbose = FALSE)

```

Plot UMAP and clusters cluster
```{r, warning=FALSE, fig.align='center'}
DimPlot(s.obj, reduction = "umap", label = TRUE, repel = TRUE) + NoLegend()
```

Add cellular detection rate
```{r}
# Switch back to 'RNA' assay
DefaultAssay(object = s.obj) <- "RNA"
# Add CDR
cdr1 <-colSums(s.obj@assays$RNA$counts)
cdr2 <-colSums(s.obj@assays$RNA$counts>0)
cdr = cdr2/cdr1
# add to metadata
s.obj$CDR = cdr
```

Plot cellular detection rate
```{r}
plot(cdr, col = alpha(labels2colors(s.obj$Group), 0.2), main="Cellular detection rate",
     ylab="CDR", pch = 19, cex = 0.3)
```


Optional, save integrated data object
```{r, warning = FALSE}
setwd(In.dir)
saveRDS(s.obj, "seurat_object_integrated.rds")
```


```{r, warning=FALSE, echo=FALSE, eval=TRUE}
setwd("C:/Users/jread/Desktop/Work/CAS_scRNAseq/New_Cleaned_data")
s.obj = readRDS("seurat_object_integrated.rds")

```


## Annotation

Cells were annotated with Azimuth, based on the human PBMC reference.

Can be run at https://azimuth.hubmapconsortium.org/

or code below


```{r, warning=FALSE, message=FALSE}
library(Azimuth)
tic()
s.obj = RunAzimuth(query = s.obj, reference = "pbmcref", verbose = TRUE)
toc()
```


Although the level 2 annotations will be used, can filter cells based on the level 1 annotation score to remove cells which have low confidence annotation at high level annotation.

```{r}
table(s.obj$predicted.celltype.l1.score > 0.5)
```

Remove low confidence cells
```{r}
s.obj = subset(x = s.obj, subset = predicted.celltype.l1.score > 0.5)
# rename
s.obj$Azimuth_2_CT = s.obj$predicted.celltype.l2
```

Cell numbers
```{r}
table(s.obj$Azimuth_2_CT)
```

Add color scheme for cell types
```{r, warning=FALSE}

set.col = c("B naive" = "royalblue2", "B intermediate" = "plum", "B memory" = "mediumseagreen",
            "CD4 Naive" = "deepskyblue", "CD4 TCM" = "coral1", "CD4 TEM" = "indianred",
            "CD4 Proliferating" = "seagreen3", "CD4 CTL" = "snow3", "Treg" = "cornflowerblue",
            "gdT" = "yellow2", "dnT" = "powderblue", "MAIT" = "palevioletred1",
            "CD8 Naive" = "darkorange", "CD8 TEM" = "navy", "CD8 TCM" = "firebrick1",
            "cDC1" = "hotpink", "cDC2" = "darkorchid1", "ASDC" = "darkorange",
            "pDC" = "deeppink3", "HSPC" = "brown3", "ILC" = "lightblue",
            "CD14 Mono" = "forestgreen", "CD16 Mono" = "gold2", "NK" = "magenta3",
            "NK_CD56bright" = "sienna3", "NK Proliferating" = "sienna2",
            "Plasmablast" = "brown", "Platelet" = "wheat3")

# add to metadata
s.obj@meta.data$Cell.Col = set.col[s.obj$Azimuth_2_CT]

```

Plot

```{r, warning=FALSE, message=FALSE}
DimPlot(s.obj, reduction = "umap", group.by = "Azimuth_2_CT", label = T,
        repel=T, cols = set.col) + NoLegend()
```

### Housekeeping

Re-arrange some variables and add color schemes

```{r}

s.obj@meta.data$Stimuli = ifelse(s.obj$Stimuli == "Unstim", "CTRL", s.obj$Stimuli)
s.obj@meta.data$Age = ifelse(s.obj$Age == "Cord", "CBMC", "5yr_PBMC")
s.obj@meta.data$Age.Stimuli_Group = paste(s.obj$Age, s.obj$Stimuli, sep = "_")


stim.col = c(CTRL = "grey80", LPS = "forestgreen", PIC = "orange2")
s.obj@meta.data$Stim_col = stim.col[as.character(s.obj$Stimuli)]

phase.col = c(G1 = "cornflowerblue", G2M = "coral2", S = "aquamarine3")
s.obj@meta.data$Phase_col = phase.col[as.character(s.obj$Phase)]

samp.col = c(Donor1 = "brown1", Donor2 = "deepskyblue2")
s.obj@meta.data$Samp_col = samp.col[as.character(s.obj$Sample)]

age.col = c("CBMC" = "cyan3", "5yr_PBMC" = "magenta3")
s.obj@meta.data$Age_col = age.col[as.character(s.obj$Age)]

set.col_G = c("CBMC_CTRL" = "khaki", "CBMC_LPS" = "green3", "CBMC_PIC" = "coral",
             "5yr_PBMC_CTRL" = "ivory3", "5yr_PBMC_LPS" = "royalblue2",
             "5yr_PBMC_PIC" = "red2")
s.obj@meta.data$Group.Col = set.col_G[as.character(s.obj$Age.Stimuli_Group)]


```

### Remove cells low numbers

```{r}
s.obj = s.obj[, (s.obj$Azimuth_2_CT == "B intermediate" | 
                    s.obj$Azimuth_2_CT == "B memory" |
                    s.obj$Azimuth_2_CT == "B naive" | 
                    s.obj$Azimuth_2_CT == "CD14 Mono" | 
                    s.obj$Azimuth_2_CT == "CD16 Mono" | 
                    s.obj$Azimuth_2_CT == "CD4 Naive" | 
                    s.obj$Azimuth_2_CT == "CD4 TCM" | 
                    s.obj$Azimuth_2_CT == "CD4 TEM" | 
                    s.obj$Azimuth_2_CT == "CD8 Naive" | 
                    s.obj$Azimuth_2_CT == "CD8 TCM" | 
                    s.obj$Azimuth_2_CT == "CD8 TEM" |
                    s.obj$Azimuth_2_CT == "dnT" |
                    s.obj$Azimuth_2_CT == "gdT" |
                    s.obj$Azimuth_2_CT == "HSPC" |
                    s.obj$Azimuth_2_CT == "ILC" |
                    s.obj$Azimuth_2_CT == "MAIT" | 
                    s.obj$Azimuth_2_CT == "NK" |
                    s.obj$Azimuth_2_CT == "NK_CD56bright" |
                    s.obj$Azimuth_2_CT == "Treg")]

table(factor(s.obj$Azimuth_2_CT))
```


### Table of cell numbers

```{r}
table(s.obj@meta.data$Azimuth_2_CT, s.obj@meta.data$Age.Stimuli_Group)
```


```{r}

x1 = as.data.frame.matrix(table(s.obj$Azimuth_2_CT, s.obj$Age.Stimuli_Group))
x1$Total = rowSums(x1)

setwd(Out.dir2)
png("1)_Celltytpe_by_group_counts.png", units="in", width=15, height=5, res=300)
pheatmap(t(x1), display_numbers = T,cluster_rows = FALSE, cluster_cols = FALSE,
         legend = TRUE, fontsize = 15, fontsize_number = 15, gaps_row = c(2, 4, 6),
         angle_col = 45, legend_breaks = c(0, 2500, 5000, 7500, 10000, 12500, 15000),
         cellheight = 25, cellwidth = 40, na_col = "grey93", labels_col = rownames(x1),
         labels_row = colnames(x1), number_format = "%.0f", number_color = "black",
         color=alpha(c("white",colorRampPalette(brewer.pal(n=4, name="Greys"))(100)),0.95),
         main = " ", cex.main = 0.6, font.main = 4)
invisible(dev.off())
```


### Plot umaps

Now plot UMAPs overlaid with variables of interest 

Cell type
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("2)_UMAP_celltype.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Azimuth_2_CT", label = F,
        repel=T, cols = set.col) + ggtitle("Azimuth define cell type annotation")
invisible(dev.off())
```

Age
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("3)_UMAP_Age.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Age",
        cols = alpha(age.col, 0.8), shuffle = TRUE, pt.size = NULL) + ggtitle("Age")
invisible(dev.off())
```

Stimuli
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("4)_UMAP_Stimuli.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Stimuli",
        cols = alpha(stim.col, 0.8), shuffle = TRUE, pt.size = NULL) + ggtitle("Stimuli")
invisible(dev.off())
```

Age and Stimuli
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("5)_UMAP_Age_Stim.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Age.Stimuli_Group",
        cols = alpha(set.col_G, 0.8), shuffle = TRUE, pt.size = NULL) +
  ggtitle("Stimuli and age")
invisible(dev.off())
```

Donor
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("6)_UMAP_Donor.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Sample",
        cols = alpha(samp.col, 0.8), shuffle = T, pt.size = NULL) +
  ggtitle("Donor")
invisible(dev.off())
```

Cell cycle phase
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)
png("7)_UMAP_Phase.png", units="in", width=6, height=5, res=300)
DimPlot(s.obj, reduction = "umap", group.by = "Phase",
        cols = alpha(phase.col, 0.8), shuffle = T, pt.size = NULL) +
  ggtitle("Cell cycle phase")
invisible(dev.off())
```


# Part 3 - Differential expression analysis


```{r}
DefaultAssay(s.obj) = "RNA"
```

#### Cord - LPS 

Subset to sample of interest
```{r}
sx = subset(s.obj, Age == "CBMC")
sx = subset(sx, Stimuli != "PIC") # excluding PolyIC samples
table(sx$Age, sx$Stimuli)
```

Choose celltypes and run the analysis
```{r, warning=FALSE, message=FALSE}

# Choose cell types to run
table(sx$Azimuth_2_CT)
ct.idx = c("B intermediate", "B naive", "CD14 Mono", "CD16 Mono", "CD4 Naive",
           "CD4 TCM", "CD8 Naive", "CD8 TEM", "gdT", "HSPC", "ILC", "NK", "Treg")

# Initiate list object to save results
CBMC_LPS_res = list()

# Loop the analysis for each cell type

for (ct in ct.idx) {
  tic()
  sx1 = subset(sx, Azimuth_2_CT == ct)
  Idents(object = sx1) = sx1$Stimuli
  # Run if bot conditiona have greater than 5 representatives
  if(table(Idents(sx1))[[1]]>5 & table(Idents(sx1))[[2]]>5) {
    DEGs <- FindMarkers(sx1, ident.1 = "LPS", ident.2 = "CTRL", test.use = "MAST",
                        latent.vars = c("CDR", "percent.mt", "Phase", "Sample"),
                        logfc.threshold = 0, min.pct = 0.1)
    # change adjP value which = 0 to the minimum
    minval = min(DEGs$p_val_adj[DEGs$p_val_adj != 0]); minval
    DEGs$p_val_adj = ifelse(DEGs$p_val_adj == 0, minval, DEGs$p_val_adj)
    # add -Log10 Padj-P value
    DEGs$nLog10_adjP = -log10(DEGs$p_val_adj)
    # Add Gene IDs
    DEGs$EnsemblID <- mapIds(org.Hs.eg.db, keys = rownames(DEGs),
                             keytype = "SYMBOL", column="ENSEMBL")
    DEGs$EntrezID <- mapIds(org.Hs.eg.db, keys = rownames(DEGs),
                            keytype = "SYMBOL", column="ENTREZID")
    DEGs$CT = ct
    DEGs$Col = set.col[DEGs$CT]
    # Add results to list
    CBMC_LPS_res[[ct]] = DEGs
    print(paste0(ct, " is done!!!"))
  }
  toc()
}

```


See example output
```{r, warning=FALSE, message=FALSE}
names(CBMC_LPS_res)
head(CBMC_LPS_res[["CD14 Mono"]])
```

### Volcano plot

Plot example volcano plot

```{r, warning=FALSE, fig.align='center'}

ct = "CD14 Mono"

DEGs = CBMC_LPS_res[[ct]]

setwd(Out.dir2)
png("8)_Example_VP_CBMC_LPS.png", units="in", width=5.5, height=5, res=300)
plot(DEGs$avg_log2FC, DEGs$nLog10_adjP, pch = 21, col = alpha("grey50", 0.5),
         bg = alpha("grey50", 0.25), xlab = "Average Log2FC",
     ylab = "-Log10 adj P value (Bonferroni)",
         main = paste(ct, "", "\nCBMC - LPS vs CTRL", sep=""),
         font.lab = 2, cex.lab=1.25, font.axis=2, cex=0.7)
    abline(h = -log10(0.01), lty=3, col = "grey50")
    points(DEGs$avg_log2FC[DEGs$p_val_adj<0.01 & DEGs$avg_log2FC>0.25],
           DEGs$nLog10_adjP[DEGs$p_val_adj<0.01 & DEGs$avg_log2FC>0.25],
           pch = 21, cex=0.7, col = alpha("firebrick1", 0.75), bg = alpha("firebrick1", 0.5))
    points(DEGs$avg_log2FC[DEGs$p_val_adj<0.01 & DEGs$avg_log2FC<(-0.25)],
           DEGs$nLog10_adjP[DEGs$p_val_adj<0.01 & DEGs$avg_log2FC<(-0.25)],
           pch = 21, cex=0.7, col = alpha("royalblue2", 0.75), bg = alpha("royalblue2", 0.5))
    invisible(dev.off())
```

### Pathways analysis

Example pathways analysis

```{r, warning=FALSE, message=FALSE, fig.align='center'}

ct = "CD14 Mono"
nam = "CBMC_LPS"

# Subset
In.dat = CBMC_LPS_res[[ct]]
# Upregulated genes
in.up = subset(In.dat, (p_val_adj < 0.01 & avg_log2FC > 0.25))
# Downregulated genes
in.down = subset(In.dat, (p_val_adj < 0.01 & avg_log2FC < -0.25))

# Choose databases for analysis
dbs <- enrichR::listEnrichrDbs()
dbs2 = dbs$libraryName[c(173, 176:178, 203)]


# Upregulated
# run enrichR and extract results
enriched <- suppressMessages(enrichr(rownames(in.up), dbs2))
r.KEGG = enriched[["KEGG_2021_Human"]]; r.KEGG$DB = "KEGG"
r.reactome = enriched[["Reactome_2022"]]; r.reactome$DB = "Reactome"
r.GOBP = enriched[["GO_Biological_Process_2021"]]; r.GOBP$DB = "GO:BP"
r.GOCC = enriched[["GO_Cellular_Component_2021"]]; r.GOCC$DB = "GO:CC"
r.GOMF = enriched[["GO_Molecular_Function_2021"]]; r.GOMF$DB = "GO:MF"
comb = rbind(r.KEGG, r.reactome, r.GOBP, r.GOCC, r.GOMF)
comb = comb[order(comb$Adjusted.P.value), ]
comb[1:10, c(1:5, 10)]
comb$colour = ifelse(comb$DB == "Reactome", "blue2",
                     ifelse(comb$DB == "KEGG", "orange",
                            ifelse(comb$DB == "GO:BP", "mediumpurple1",
                                   ifelse(comb$DB == "GO:MF", "darkseagreen2", "indianred1"))))

# Optional: save results
#write.csv(comb, paste0(ct, "_UP_Pathways.csv"))

# Subset to top 25 results
comb1 = comb[1:25, ]
# Shorten names for plotting
nam.ls = ifelse(nchar(comb1$Term) > 50,
                paste(substr(comb1$Term, 1, 50), "...", sep=""), comb1$Term)
# plot

setwd(Out.dir2)
png("9)_Example_Pathways_CBMC_LPS.png", units="in", width=12, height=10, res=300)
par(mar = c(5.1, 25.1, 4.1, 2.1))
barplot(rev(-log10(comb1$Adjusted.P.value)), horiz = T, xlab = "-Log10(Adjused-P Value)",
        main = paste0(nam, " - ", ct, "\n", "Enriched pathways from upregulated genes"),
        col = alpha(rev(comb1$colour), 0.5), border = rev(comb1$colour), space = 0.5,
        names = rev(nam.ls), las = 2, xaxt="n", cex.lab = 1.2, font.lab = 2,font = 2, cex.main = 1.2, cex.names = 1.0)
axis(1, at = 1:round(max(-log10(comb1$Adjusted.P.value)),0), 
     labels = 1:round(max(-log10(comb1$Adjusted.P.value)),0), las=1)
abline(v=-log10(0.001), col = 'red', lty=2, lwd=2)
legend("bottomright", legend = c("Reactome", "KEGG", "GO:BP", "GO:MF", "GO:CC"), 
       col = c("blue2", "orange", "mediumpurple1", "darkseagreen2", "indianred1"),
       pt.bg = alpha(c("blue2", "orange", "mediumpurple1", "darkseagreen2", "indianred1"), 0.5),
       pch = 22, pt.cex=4, cex=1.75)
invisible(dev.off())
```

### Multiplot

```{r, warning=FALSE, message=FALSE}
# Bind results
CL_fp = rbind(CBMC_LPS_res[["B intermediate"]], CBMC_LPS_res[["B naive"]],
              CBMC_LPS_res[["CD14 Mono"]], CBMC_LPS_res[["CD16 Mono"]],
              CBMC_LPS_res[["CD4 Naive"]], CBMC_LPS_res[["CD4 TCM"]],
              CBMC_LPS_res[["CD8 Naive"]], CBMC_LPS_res[["CD8 TEM"]],
              CBMC_LPS_res[["gdT"]], CBMC_LPS_res[["HSPC"]],
              CBMC_LPS_res[["ILC"]], CBMC_LPS_res[["NK"]],
              CBMC_LPS_res[["Treg"]])
# subset to DEGs
CL_fp = subset(CL_fp, (avg_log2FC > 0.25 | avg_log2FC < -0.25))
CL_fp = subset(CL_fp, p_val_adj < 0.01)
# order
CL_fp1 = CL_fp[order(CL_fp$CT, -CL_fp$p_val_adj), ]
table(CL_fp1$CT)
# index
CL_fp1$IDX = 1:nrow(CL_fp1)
```

Plot
```{r, warning=FALSE, message=FALSE}
setwd(Out.dir2)

png("10)_CBMC_LPS_Summary_VP.png", units="in", width=9, height=5, res=600)
par(mar = c(3.1, 4.1, 4.5, 4.1) )
plot(CL_fp1$IDX, CL_fp1$nLog10_adjP, xlab="", ylab="",  
     axes=FALSE, type="l", col=alpha("grey70", 80), lty = 3)
abline(h=2, col = "black", lty=1, lwd=1)
mtext("-Log10 adj-P value",side=4,col="black",line=2.9, las = 0, cex=1.25) 
axis(4, col="black",col.axis="black",las=1)
par(new=TRUE)
plot(CL_fp1$IDX, CL_fp1$avg_log2FC, cex = 0.5, lwd = 0.5, pch =21,
     col = alpha(CL_fp1$Col, 0.4),
     main = "", xlab = "", ylab = "", xaxt = 'n', cex.main=0.9,
     bg = alpha(CL_fp1$Col, 0.2))
title("CBMC - LPS vs CTRL\nDifferentially expressed genes\ncompared to control",
      cex.main = 1.1, line = 0.5, adj = 0.0, font.main = 3)
title(ylab="Average Log2FC", line=2.5, cex.lab=1.25)
title(xlab="Decreasing adjusted-P value (stratified by cell type)
      (minimum adjusted-p value < 0.01)",
      line=1.3, cex.lab=1.2)
abline(h = -0.25, col = alpha("blue2", 0.5), lty=1, lwd=1)
abline(h = 0.25, col = alpha("red2", 0.5), lty=1, lwd=1)
# Add gene label to plot
degs = CL_fp1[(CL_fp1$avg_log2FC<(-4) | CL_fp1$avg_log2FC> 4), ] # Change to highlight genes
addTextLabels(degs$IDX, degs$avg_log2FC, labels = rownames(degs), avoidPoints = T,
              cex.label=0.7, col.label=alpha(degs$Col, 0.5), col.line = alpha(degs$Col, 0.3))
# Add legend
legend("topright",
       legend = c("B naive", "B intermediate",
                  "CD4 Naive", "CD4 TCM", "Treg",
                  "CD8 Naive", "CD8 TEM", "gdT",
                  "NK", "CD14 mono", "CD16 mono",
                  "HSPC"), 
       col = c("royalblue2", "plum", 
               "deepskyblue", "coral1", "cornflowerblue",
               "darkorange", "navy", "yellow2",
               "magenta3", "forestgreen", "gold2",
               "brown3"),
       pt.bg = alpha(c("royalblue2", "plum", 
                       "deepskyblue", "coral1", "cornflowerblue",
                       "darkorange", "navy", "yellow2",
                       "magenta3", "forestgreen", "gold2",
                       "brown3"), 0.5),
       pch=21, pt.cex=1.1, cex=0.65, ncol=4, inset=c(0,-0.15), xpd=TRUE)
invisible(dev.off())
```



# Part 4 - Monocle analysis

To be added

# Part 5 - CellCall analysis

To be added

# Session info

```{r, warning=FALSE}
sessionInfo()
```

